<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Generate count (From RNAseq BAM files) &mdash; Renjin Benchmarks 0.1 documentation</title>
    
    <link rel="stylesheet" href="../../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Renjin Benchmarks 0.1 documentation" href="../../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../../index.html">Renjin Benchmarks 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="generate-count-from-rnaseq-bam-files">
<h1>Generate count (From RNAseq BAM files)<a class="headerlink" href="#generate-count-from-rnaseq-bam-files" title="Permalink to this headline">¶</a></h1>
<p>The amount of RNA produced by a specific gene is used as a surrogate marker for detection the activity, of
that gene. With Current highthroughput sequencing technologies (such as in this case RNAseq), all (fragmented)
RNA molecules from a biological sample are sequenced. These sequences are then matched to the annotated genome
sequence to identify to which gene each sequenced fragment belongs. More sequenced fragments mapping to a gene
in one samples as compared to another, means that the specific gene had a higher activity.</p>
<p>The first step in this workflow is reading the relevant parts of genome annotation file and converting it the
a list of ranges of feature of interest such as exon (IRanges, GenomicRanges). Sequencing products previously
aligned to genome are imported using Rsamtools. Mapping of mapped sequences to genes as defined in the anotation
file is the next step. Followed by aggragation of data by summing the number of mapped fragments per gene.</p>
<div class="section" id="packages-and-dependencies">
<h2>Packages and Dependencies<a class="headerlink" href="#packages-and-dependencies" title="Permalink to this headline">¶</a></h2>
<p>There are 5 packages used in this workflow, which depend on 6 additional packages from Bioconductor and
CRAN (dependencies)</p>
<p><strong>Used packages:</strong></p>
<ul class="simple">
<li><strong>Bioconductor</strong>: BiocGenerics, GenomicRanges, IRanges, Rsamtools</li>
</ul>
<ul class="simple">
<li><em>CRAN</em>: parallel</li>
</ul>
<p><strong>Package dependencies:</strong></p>
<ul class="simple">
<li><em>Bioconductor</em>: Biostrings, GenomeInfoDb, XVector, S4Vectors, zlibbioc</li>
<li><em>CRAN</em>: bitops</li>
</ul>
</div>
<div class="section" id="data">
<h2>Data:<a class="headerlink" href="#data" title="Permalink to this headline">¶</a></h2>
<p>Bam files are single-end RNAseq data from Kartalaei PS et al, 2015.</p>
</div>
<div class="section" id="license">
<h2>License<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Copyright (c) 2015-2016 BeDataDriven B.V.</div>
<div class="line">License: <a class="reference external" href="http://www.apache.org/licenses/LICENSE-2.0">Apache License version 2.0 or higher</a></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Generate count (From RNAseq BAM files)</a><ul>
<li><a class="reference internal" href="#packages-and-dependencies">Packages and Dependencies</a></li>
<li><a class="reference internal" href="#data">Data:</a></li>
<li><a class="reference internal" href="#license">License</a></li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../../_sources/bioinformatics/generate_count/README.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../../index.html">Renjin Benchmarks 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2016, BeDataDriven B.V..
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>