<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Integration: iGraph &#8212; Renjin Benchmarks 0.1 documentation</title>
    
    <link rel="stylesheet" href="../../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="top" title="Renjin Benchmarks 0.1 documentation" href="../../index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="../../index.html">Renjin Benchmarks 0.1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="integration-igraph">
<h1>Integration: iGraph<a class="headerlink" href="#integration-igraph" title="Permalink to this headline">¶</a></h1>
<p>Genes are sequences of DNA that encode functional molecules (such as RNAs or Proteins).
These functional molecules most of time interact and affect eachothers function.
Therefor, to understand the role of a specific gene in a disease, scientist need to
know about all the published functions.</p>
<p>Given many published scientific manuscripts about any given gene, the <a class="reference external" href="http://www.ncbi.nlm.nih.gov">National
Center for Biotechnology Information (NCBI)</a> has introduced <a class="reference external" href="http://www.ncbi.nlm.nih.gov/gene/about-generif">GeneRIF</a>. GeneRIF
provides scientist a simple mechanism to add functional annotation to known genes
based on scientific publications. These annotations are than processed and approved
by GeneRIF staff. All these functional annotations are freely accessible and can be
used by anyone to extract functions of a gene or genes with specific function.</p>
<p>This workflow imports human Phophatases and Kinases and the publications. And uses
&#8216;igraph&#8217; to store this information and to plot the relation between these publications
as well as interaction network between molecules.</p>
<div class="figure" id="id1">
<img src="../../_images/graphviz-b1c82a2746f58544e0cf30092186d2bce684f013.png" alt="digraph INTEGRATE_liver {
             compound=true;

             print [label = &quot;print()&quot;];

             data_pk [shape = invhouse, label = &quot;human PPases &amp;  \nkinases from GO&quot;];
             data_sql [shape = invhouse, label = &quot;SQL query  \nget_all_RIF.sql&quot;];
             data_rif [shape = invhouse, label = &quot;RIFs \nfrom entrez&quot;];
             read [shape = box, label = &quot;Read\ngunzip()\nread.delim()&quot;];
             edges [shape = box, label = &quot;Extract edges\nsqldf()&quot;];
             network [shape = box; label = &quot;Make graph object\ngraph.data.frame()   &quot;];
             subset [label = &quot;subset&quot;];
             network2 [shape = box; label = &quot;Make graph object\ngraph.data.frame()   &quot;];
             decomp [shape = box; label = &quot;Decompose\ndecompose.graph()   &quot;];
             largest [label = &quot;select largest     \ncomponent&quot;]
             vertices [shape = box; label = &quot;extract vertices \nget.data.frame()   &quot;];
             cocitation [shape = box; label = &quot;peform cocitation\ngraph.adjacency(cocitation())        &quot;];
             subgrph [shape = box; label = &quot;subset gene nodes\ninduced.subgraph()&quot;];
             nodeprop [label = &quot;change node    \nproperties&quot;];
             getdata [label = &quot;get.data.frame()&quot;];

             subgraph cluster_0 {
             label = &quot;Load data&quot;;
                     style = filled;
                     color = lightgrey;
                     data_rif -&gt; read;
                     read -&gt; query;
                     query -&gt; edges;
                     edges -&gt; network;
                     data_sql -&gt; edges;
             }
             subgraph cluster_1 {
             label = &quot;Decompose&quot;;
                     style = filled;
                     color = lightgrey;
                     node [style = filled, color = white];
                     decomp -&gt; largest;
             }
             subgraph cluster_2 {
             label = &quot;Do cocitation&quot;;
                     style = filled;
                     color = lightgrey;
                     node [style = filled, color = white];
                     vertices -&gt; cocitation;
                     cocitation -&gt; subgrph;
             }

             network -&gt; subset [ltail = cluster_0];
             data_pk -&gt; subset -&gt; network2;
             network2 -&gt; decomp [lhead = cluster_1];
             largest -&gt; vertices [ltail = cluster_1, lhead = cluster_2];
             subgrph-&gt; nodeprop [ltail = cluster_2];
             nodeprop -&gt; getdata -&gt; print;

     { rank=same;
     network -&gt; subgrph [style=invis];
     }

             # do.load: data_rif-&gt;read-&gt;query-&gt;edges-&gt;network;data_sql-&gt;edges;
             # do.mesh
             terms [shape = invhouse; label = &quot;Ontology terms&quot;];
             pubmed [shape = box; label = &quot;Query PubMed&quot;];
             listxml [shape = box; label = &quot;xmlToList(xmlParse())&quot;];
             network3 [shape = box; label = &quot;Make graph object\ngraph.data.frame()   &quot;];
             filter [label = &quot;Filter&quot;];
             terms -&gt; pubmed -&gt; listxml -&gt; filter -&gt; network3;
             edges -&gt; filter [ltail = cluster_0];
             network3 -&gt; decomp [lhead = cluster_1];
     }" />
<p class="caption"><span class="caption-text">Workflow for integration liver cohort data.</span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</div>
<div class="section" id="packages-and-dependencies">
<h2>Packages and Dependencies<a class="headerlink" href="#packages-and-dependencies" title="Permalink to this headline">¶</a></h2>
<p>There are 7 packages used in this workflow, which depend on 35 additional packages
from CRAN (dependencies)</p>
<p><strong>Used packages:</strong></p>
<ul class="simple">
<li><em>CRAN</em>: igraph, XML, R.utils, plyr, reshape, utils, sqldf</li>
</ul>
<p><strong>Package dependencies:</strong></p>
<ul class="simple">
<li><em>CRAN</em>: magrittr, irlba, Matrix, NMF, lattice, foreach, gridBase, pkgmaker, reshape2, stringr, colorspace, doParallel, digest, rngtools, ggplot2, RColorBrewer, cluster, registry, codetools, iterators, xtable, plyr, Rcpp, stringi, scales, gtable, MASS, proto, dichromat, labeling, munsell, RSQLite, gsubfn, chron, DBI</li>
</ul>
</div>
<div class="section" id="data">
<h2>Data<a class="headerlink" href="#data" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Basic GeneRIF data:
<a class="reference external" href="ftp://ftp.ncbi.nih.gov/gene/GeneRIF/generifs_basic.gz">ftp://ftp.ncbi.nih.gov/gene/GeneRIF/generifs_basic.gz</a></li>
<li>SQL script written by Ieuan Clay:
get_all_RIF.sql</li>
<li>Human kinases and phosphatases protein:
GeneOntology.org list of human (Taxonomy id: 9606) genes using terms <a class="reference external" href="GO:0050222">GO:0050222</a>
(protein kinase activity) and <a class="reference external" href="GO:0004721">GO:0004721</a> (phosphoprotein phosphatase activity).</li>
</ul>
</div>
<div class="section" id="license">
<h2>License<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Copyright (c) 2015 Ieuan Clay based on code from <a class="reference external" href="https://github.com/biolion/genbench">genbench</a></li>
<li>Copyright (c) 2015-2016 BeDataDriven B.V.  License: <a class="reference external" href="http://www.gnu.org/licenses/gpl.html">GPL version 2 or higher</a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Integration: iGraph</a><ul>
<li><a class="reference internal" href="#packages-and-dependencies">Packages and Dependencies</a></li>
<li><a class="reference internal" href="#data">Data</a></li>
<li><a class="reference internal" href="#license">License</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../../_sources/bioinformatics/integration_igraph/README.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="../../index.html">Renjin Benchmarks 0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2016, BeDataDriven B.V..
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.8.
    </div>
  </body>
</html>